s3/pkg/datastore/yig/config/watcher.go

```diff
@@ -0,0 +1,97 @@
+package config
+
+import (
+	"errors"
+	"strings"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+)
+
+const (
+	CFG_SUFFIX = "toml"
+)
+
+type ConfigWatcher struct {
+	FuncConfigParse FuncConfigParse
+	watcher         *fsnotify.Watcher
+	stopSignal      chan bool
+	wg              sync.WaitGroup
+	err             error
+}
+
+func (cw *ConfigWatcher) Error() error {
+	return cw.err
+}
+
+func (cw *ConfigWatcher) Stop() {
+	cw.stopSignal <- true
+	close(cw.stopSignal)
+	cw.wg.Wait()
+	cw.watcher.Close()
+}
+
+func (cw *ConfigWatcher) Watch(dir string) {
+	mask := fsnotify.Write | fsnotify.Create
+	cw.wg.Add(1)
+	go func() {
+		defer cw.wg.Done()
+		for {
+			select {
+			case event, ok := <-cw.watcher.Events:
+				if !ok {
+					cw.err = errors.New("failed to read watcher events.")
+					return
+				}
+				if event.Op&mask != 0 {
+					// got we need.
+					if strings.HasSuffix(event.Name, CFG_SUFFIX) {
```

Comment:
how does this interplay with all the structs defined above(for various configs)?